tests/orion/orchestration/test_core_policy.py

```diff
@@ -1553,3 +1552,321 @@ async def test_concurrency_race_condition_limit_decreases_dont_impact_existing_r
+                mutated_state = proposed_state.copy()
+                mutated_state.type = random.choice(
+                    list(
+                        set(states.StateType)
+                        - {initial_state.type, proposed_state.type}
+                    )
+                )
+                await self.reject_transition(
+                    mutated_state, reason="gotta fizzle some rules, for fun"
+                )
+
+            async def after_transition(self, initial_state, validated_state, context):
+                pass
+
+            async def cleanup(self, initial_state, validated_state, context):
+                pass
+
+        accept_concurrency_policy = [
+            SecureTaskConcurrencySlots,
+            ReleaseTaskConcurrencySlots,
+        ]
+
+        reject_concurrency_policy = [
+            ReleaseTaskConcurrencySlots,
+            SecureTaskConcurrencySlots,
+            StateMutatingRule,
+        ]
+
+        await self.create_concurrency_limit(session, "small", 1)
+
+        # Fill the concurrency slot by transitioning into a running state
+        running_transition = (states.StateType.PENDING, states.StateType.RUNNING)
+
+        ctx = await initialize_orchestration(
+            session, "task", *running_transition, run_tags=["small"]
+        )
+
+        async with contextlib.AsyncExitStack() as stack:
+            for rule in accept_concurrency_policy:
+                task1_running_ctx = await stack.enter_async_context(
+                    rule(ctx, *running_transition)
+                )
+            await task1_running_ctx.validate_proposed_state()
+
+        assert task1_running_ctx.response_status == SetStateStatus.ACCEPT
+        assert (await self.count_concurrency_slots(session, "small")) == 1
+
+        # Make sure that the concurrency slot is released even though the transition was
+        # rejected, because the task was still moved out of a RUNNING state
+        pending_transition = (states.StateType.RUNNING, states.StateType.PENDING)
+
+        task1_pending_ctx = await initialize_orchestration(
+            session,
+            "task",
+            *pending_transition,
+            run_override=task1_running_ctx.run,
+            run_tags=["small"],
+        )
+        async with contextlib.AsyncExitStack() as stack:
+            for rule in reject_concurrency_policy:
+                task1_pending_ctx = await stack.enter_async_context(
+                    rule(task1_pending_ctx, *pending_transition)
+                )
+            await task1_pending_ctx.validate_proposed_state()
+
+        assert task1_pending_ctx.response_status == SetStateStatus.REJECT
+        assert task1_pending_ctx.validated_state.type != states.StateType.RUNNING
+        assert (await self.count_concurrency_slots(session, "small")) == 0
+
+    async def test_returning_concurrency_slots_when_transitioning_out_of_running_even_on_fizzle_with_release_last(
```

Comment:
for this test, we aren't "fizzling" the concurrency slot release, we'd be invalidating it instead (though of course these concepts don't really apply to transforms, only rules)